// z-secure/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PostgresTokenBucketStore = exports.RedisFixedWindowStore = exports.RedisTokenBucketStore = exports.MemoryTokenBucketStore = exports.MemoryLeakyBucketStore = exports.MemoryFixedWindowStore = exports.tokenBucket = exports.FixedWindow = exports.LeakyBucket = exports.PostgresLeakyBucketStore = exports.PostgresFixedWindowStore = exports.PostgresShieldStore = exports.RedisShieldStore = exports.ShieldMemoryStore = exports.ZShield = void 0;
const pg_1 = __importDefault(require("./fixed-window/pg"));
exports.PostgresFixedWindowStore = pg_1.default;
const pg_2 = __importDefault(require("./leaky-bucket/pg"));
exports.PostgresLeakyBucketStore = pg_2.default;
const lib_lb_1 = __importDefault(require("./leaky-bucket/lib-lb"));
exports.LeakyBucket = lib_lb_1.default;
const lib_fixed_window_1 = __importDefault(require("./fixed-window/lib-fixed-window"));
exports.FixedWindow = lib_fixed_window_1.default;
const lib_1 = __importDefault(require("./token-bucket/lib"));
exports.tokenBucket = lib_1.default;
const memory_fw_1 = __importDefault(require("./fixed-window/memory-fw"));
exports.MemoryFixedWindowStore = memory_fw_1.default;
const memory_lb_1 = __importDefault(require("./leaky-bucket/memory-lb"));
exports.MemoryLeakyBucketStore = memory_lb_1.default;
const memory_1 = __importDefault(require("./token-bucket/memory"));
exports.MemoryTokenBucketStore = memory_1.default;
const cache_memory_1 = __importDefault(require("./token-bucket/cache-memory"));
exports.RedisTokenBucketStore = cache_memory_1.default;
const cache_memory_2 = __importDefault(require("./fixed-window/cache-memory"));
exports.RedisFixedWindowStore = cache_memory_2.default;
const lib_2 = __importDefault(require("./shield/lib"));
exports.ZShield = lib_2.default;
const cache_memory_3 = __importDefault(require("./shield/memory/cache-memory"));
exports.RedisShieldStore = cache_memory_3.default;
const pg_3 = __importDefault(require("./shield/memory/pg"));
exports.PostgresShieldStore = pg_3.default;
const inMemoryStore_1 = __importDefault(require("./shield/memory/inMemoryStore"));
exports.ShieldMemoryStore = inMemoryStore_1.default;
const pg_4 = __importDefault(require("./token-bucket/pg"));
exports.PostgresTokenBucketStore = pg_4.default;
__exportStar(require("./types"), exports);
